vision_doc.md

```diff
@@ -0,0 +1,279 @@
+    ///   inherent `builder` function on the target type.
+    #[derive(Debug, Clone, Builder, Serialize)]
+    #[builder(setter(into), try_setter, preserve_attrs(serde), derive = "Deserialize", build_fn(skip), target_factory)]
+    pub struct StatRequest {
+        /// This particular field should be named `stat` on the inbound request;
+        /// that requires a field-level annotation.
+        #[serde(rename(deserialize = "stat"))]
+        stat_name: String,
+
+        cycle: Cycle,
+
+        #[serde(default, rename = "filter", skip_serializing_if="Option::is_none")]
+        #[builder(default)]
+        key_filter: Option<Filter>,
+    }
+
+    impl StatRequest {
+        pub fn new(stat_name: String, cycle: Cycle) -> Result<Self, String> {
+            StatRequest::with_filter(stat_name, cycle, None)
+        }
+
+        /// Creates a new `StatRequest` with the specified filter (`None` is allowed).
+        ///
+        /// A filter is required for detail stats for perf reasons; this method will 
+        /// return an error if one is needed and `None` is provided, or if one is provided
+        /// and the specified stat does not support filtering.
+        pub fn with_filter<F: Into<Option<Filter>>>(stat_name: String,
+                                                   cycle: Cycle,
+                                                   filter: F)
+                                                   -> Result<Self, String> {
+            let requires_filter = stat_name.ends_with("_detail");
+            let filter_opt = filter.into();
+            match (requires_filter, filter_opt.is_some()) {
+                (true, false) => Err(format!("Stat `{}` requires a filter", stat_name)),
+                (false, true) => Err(format!("Stat `{}` is not a detail metric")),
+                _ => {
+                    Ok(StatRequest {
+                        stat_name: stat_name,
+                        key_filter: filter_opt,
+                        cycle: cycle,
+                    })
+                }
+            }
+        }
+    }
+    
+    impl StatRequestBuilder {
+        /// Create a stat request by calling `StatRequest::with_filter`. This method
+        /// will return an error if a required field is not initialized, or if the 
+        /// values fail validation.
+        pub fn build(&self) -> Result<Self, String> {
+            // These getters are private methods which return Result<T, String>;
+            // the error arises if the field is uninitialized and has no default.
+            StatRequest::with_filter(
+                self.get_stat_name()?, 
+                self.get_cycle()?, 
+                self.get_filter()?)
+        }
+    }
+    
+    impl TryFrom<StatRequestBuilder> for StatRequest {
+        type Error = String;
+        
+        fn try_from(v: StatRequestBuilder) -> Result<Self, Self::Error> {
+            v.build()
+        }
+    }
```

Comment:
The `TryFrom` could be automatically generated if the user told us that the build method was implemented and what its name was, but I take your point. I was also trying to avoid any magic that depended on field order being the same in the struct declaration and the constructor, although something could be done with that which would make the code shorter.

I tried to pick an example where the validation logic was non-trivial and would need to be applied regardless of how the object was assembled.